azure-iot-device/tests/iothub/pipeline/test_mqtt_topic_iothub.py

```diff
@@ -0,0 +1,1014 @@
+    @pytest.mark.parametrize(
+        "topic",
+        [
+            pytest.param("not a topic", id="Not a topic"),
+            pytest.param(
+                "devices/fake_device/modules/fake_module/inputs/fake_input",
+                id="Topic of wrong type",
+            ),
+            pytest.param("$iothub/methdos/POST/fake_method/?$rid=1", id="Malformed topic"),
+        ],
+    )
+    def test_invalid_method_topic(self, topic):
+        with pytest.raises(ValueError):
+            mqtt_topic_iothub.get_method_request_id_from_topic(topic)
+
+
+@pytest.mark.describe(".get_twin_request_id_from_topic()")
+class TestGetTwinRequestIdFromTopic(object):
+    @pytest.mark.it("Returns the request id from a twin response topic")
+    def test_valid_twin_response_topic(self):
+        topic = "$iothub/twin/res/200/?rid=1"
+        expected_request_id = "1"
+
+        assert mqtt_topic_iothub.get_twin_request_id_from_topic(topic) == expected_request_id
+
+    # NOTE: valid request ids shouldn't need to be URL decoded, but we do it for safety
+    # and consistency. As a result, this test covers request_id values that are not valid
+    @pytest.mark.it("URL decodes the returned value")
+    def test_url_decodes_value(self):
+        topic = "$iothub/twin/res/200/?rid=fake%24request%2Fid"
+        expected_request_id = "fake$request/id"
+        assert mqtt_topic_iothub.get_twin_request_id_from_topic(topic) == expected_request_id
+
+    @pytest.mark.it("Raises a ValueError if the provided topic is not a twin response topic")
+    @pytest.mark.parametrize(
+        "topic",
+        [
+            pytest.param("not a topic", id="Not a topic"),
+            pytest.param(
+                "devices/fake_device/modules/fake_module/inputs/fake_input",
+                id="Topic of wrong type",
+            ),
+            pytest.param("$iothub/twn/res/200?rid=1", id="Malformed topic"),
+        ],
+    )
+    def test_invalid_twin_response_topic(self, topic):
+        with pytest.raises(ValueError):
+            mqtt_topic_iothub.get_twin_request_id_from_topic(topic)
+
+
+@pytest.mark.describe(".get_twin_status_code_from_topic()")
+class TestGetTwinStatusCodeFromTopic(object):
+    @pytest.mark.it("Returns the status from a twin response topic")
+    def test_valid_twin_response_topic(self):
+        topic = "$iothub/twin/res/200/?rid=1"
+        expected_status = "200"
+
+        assert mqtt_topic_iothub.get_twin_status_code_from_topic(topic) == expected_status
+
+    @pytest.mark.it("Raises a ValueError if the provided topic is not a twin response topic")
+    @pytest.mark.parametrize(
+        "topic",
+        [
+            pytest.param("not a topic", id="Not a topic"),
+            pytest.param(
+                "devices/fake_device/modules/fake_module/inputs/fake_input",
+                id="Topic of wrong type",
+            ),
+            pytest.param("$iothub/twn/res/200?rid=1", id="Malformed topic"),
```

Comment:
>Malformed [](start = 58, length = 9)

OK, I can see why this one is malformed.  Maybe add how it's malformed to the id (e.g. "Malformed topic (missing final slash)").  Here and everywhere.

Reply:
It's a typo of "twin" as "twn"

---
In reply to: [399807979](https://github.com/Azure/azure-iot-sdk-python/pull/514#discussion_r399807979) [](ancestors = 399807979)

Reply:
For now I'm going to pass on making these more detailed, because it opens the door to testing tons of different types of malformed topics.... which are quite honestly not really valid, because we receive them from service.

It also will then get into how poorly we really check malformed topics at present.

If this ever becomes relevant, we can expand it then, until such time, I think it's best to just keep it simple

---
In reply to: [400405654](https://github.com/Azure/azure-iot-sdk-python/pull/514#discussion_r400405654) [](ancestors = 400405654,399807979)